python/taichi/lang/transformer.py

```diff
@@ -16,25 +17,88 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         local = self.t.local_scopes[-1]
         if self.stmt_block is not None:
             for var in reversed(local):
-                stmt = ASTTransformer.parse_stmt('del var')
+                stmt = ASTTransformerBase.parse_stmt('del var')
                 stmt.targets[0].id = var
                 self.stmt_block.append(stmt)
         self.t.local_scopes = self.t.local_scopes[:-1]
 
 
-# Single-pass transform
-class ASTTransformer(ast.NodeTransformer):
+# Total transform
+# TODO: ASTTransformerBase -> ASTTransformer
+# TODO: ASTTransformer -> ASTTransformerTotal
```

Comment:
OK, let's refactor kernel&func transforming procedual iapr.

Reply:
> So in the TODO, instead of naming this to `ASTTransformerTotal`, we can just wrap these passes a simple function. (Ideally this can be something like a pass manager, but we are far from that goal.)

Agreed!